File: app/src/main/java/com/cajetan/youtubeplus/PlayerActivity.java
package com.cajetan.youtubeplus;

import android.app.DialogFragment;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.arch.lifecycle.ViewModelProviders;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;
import android.support.v4.media.session.MediaButtonReceiver;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import com.cajetan.youtubeplus.data.VideoData;
import com.cajetan.youtubeplus.data.VideoDataViewModel;
import com.cajetan.youtubeplus.utils.FullScreenHelper;
import com.cajetan.youtubeplus.utils.YouTubeData;
import com.google.api.services.youtube.model.Video;
import com.pierfrancescosoffritti.androidyoutubeplayer.player.PlayerConstants;
import com.pierfrancescosoffritti.androidyoutubeplayer.player.YouTubePlayer;
import com.pierfrancescosoffritti.androidyoutubeplayer.player.YouTubePlayerView;
import com.pierfrancescosoffritti.androidyoutubeplayer.player.listeners.AbstractYouTubePlayerListener;
import com.pierfrancescosoffritti.androidyoutubeplayer.player.listeners.YouTubePlayerFullScreenListener;
import com.pierfrancescosoffritti.androidyoutubeplayer.player.listeners.YouTubePlayerInitListener;
import com.pierfrancescosoffritti.androidyoutubeplayer.ui.PlayerUIController;
import com.pierfrancescosoffritti.androidyoutubeplayer.ui.menu.MenuItem;
import com.pierfrancescosoffritti.androidyoutubeplayer.utils.YouTubePlayerTracker;

import java.io.IOException;
import java.net.URL;

public class PlayerActivity extends AppCompatActivity
        implements YouTubeData.VideoDataListener, SeekDialog.SeekDialogListener {

    private static final String TAG = PlayerActivity.class.getSimpleName();

    private FullScreenHelper fullScreenHelper = new FullScreenHelper(this);

    private YouTubePlayerView mainPlayerView;

    private static MediaSessionCompat mMediaSession;
    private PlaybackStateCompat.Builder mStateBuilder;

    private YouTubeData youTubeData;
    private PlayerUIController mUIController;

    private String mVideoId;
    private Video mVideoData;
    private Bitmap mVideoThumbnail;

    private Context mContext;

    private VideoDataViewModel mVideoDataViewModel;

    private final YouTubePlayerTracker mTracker = new YouTubePlayerTracker();

    // TODO: fetch youtube data after restoring internet connection

    /*//////////////////////////////////////////////////////////////////////////////
    // Lifecycle
    //////////////////////////////////////////////////////////////////////////////*/

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_player);

        mContext = this;
        mVideoId = getIntentVideoId();

        youTubeData = new YouTubeData(this);
        youTubeData.receiveVideoData(mVideoId);

        setupPlayer();
        setupMediaSession();

        mVideoDataViewModel = ViewModelProviders.of(this).get(VideoDataViewModel.class);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        Log.d(TAG, "Getting a new intent");
        Log.d(TAG, "New video id: " + getIntentVideoId());
        super.onNewIntent(intent);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        youTubeData.onParentActivityResult(requestCode, resultCode, data);
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        youTubeData.receiveVideoData(mVideoId);
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    protected void onDestroy() {
        ((NotificationManager) getSystemService(NOTIFICATION_SERVICE)).cancel(0);
        mainPlayerView.release();
        super.onDestroy();
    }

    /*//////////////////////////////////////////////////////////////////////////////
    // Init
    //////////////////////////////////////////////////////////////////////////////*/

    private void setupPlayer() {
        startService(new Intent(this, PlayerLifecycleService.class));

        final String videoId = mVideoId;
        mainPlayerView = findViewById(R.id.main_player_view);
        mainPlayerView.enableBackgroundPlayback(true);
        mainPlayerView.initialize(new YouTubePlayerInitListener() {
            @Override
            public void onInitSuccess(@NonNull final YouTubePlayer initialisedYouTubePlayer) {
                initialisedYouTubePlayer.addListener(mTracker);

                initialisedYouTubePlayer.addListener(new AbstractYouTubePlayerListener() {
                    @Override
                    public void onReady() {
                        initialisedYouTubePlayer.loadVideo(videoId, 0);
                    }

                    @Override
                    public void onStateChange(@NonNull PlayerConstants.PlayerState state) {
                        // TODO: not sure if it's a good idea to cast tbh

                        if (state == PlayerConstants.PlayerState.PLAYING) {
                            mStateBuilder.setState(PlaybackStateCompat.STATE_PLAYING,
                                    (long) mTracker.getCurrentSecond(), 1f);
                        } else if (state == PlayerConstants.PlayerState.PAUSED) {
                            mStateBuilder.setState(PlaybackStateCompat.STATE_PAUSED,
                                    (long) mTracker.getCurrentSecond(), 1f);
                        }

                        mMediaSession.setPlaybackState(mStateBuilder.build());
                        showMediaNotification(mStateBuilder.build());

                        super.onStateChange(state);
                    }
                });
            }
        }, true);

        mainPlayerView.addFullScreenListener(new YouTubePlayerFullScreenListener() {
            @Override
            public void onYouTubePlayerEnterFullScreen() {
                setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
                fullScreenHelper.enterFullScreen();
            }

            @Override
            public void onYouTubePlayerExitFullScreen() {
                setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
                fullScreenHelper.exitFullScreen();
            }
        });

        mUIController = mainPlayerView.getPlayerUIController();

        if (mUIController.getMenu() == null)
            return;

        mUIController.showMenuButton(true);
        mUIController.getMenu().addItem(new MenuItem("Seek", R.drawable.ic_timer_black_24dp, new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Bundle b = new Bundle();
                b.putString("duration_string", mVideoData.getContentDetails().getDuration());
                b.putInt("current_second", Math.round(mTracker.getCurrentSecond()));

                DialogFragment newFragment = new SeekDialog();
                newFragment.setArguments(b);
                newFragment.show(getFragmentManager(), "seeker_dialog");

                mUIController.getMenu().dismiss();
            }
        }));

        // TODO: option to unstar if already starred

        mUIController.getMenu().addItem(new MenuItem("Add to favourites",
                R.drawable.ic_star_border_black_24dp, new View.OnClickListener() {

            @Override
            public void onClick(View v) {
                Log.d("PlayerActivity", "Added a video to favourites");
                mVideoDataViewModel.insert(new VideoData(mVideoId));

                Toast.makeText(mContext, "Added to favourites", Toast.LENGTH_SHORT).show();

                mUIController.getMenu().dismiss();
            }

        }));
    }

    private void setupMediaSession() {
        mMediaSession = new MediaSessionCompat(this, TAG);

        mMediaSession.setFlags(
                MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS |
                        MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS);

        mMediaSession.setMediaButtonReceiver(null);

        mStateBuilder = new PlaybackStateCompat.Builder()
                .setActions(
                        PlaybackStateCompat.ACTION_PLAY |
                                PlaybackStateCompat.ACTION_PAUSE |
//                                PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS |
                                PlaybackStateCompat.ACTION_PLAY_PAUSE);

        mMediaSession.setPlaybackState(mStateBuilder.build());
        mMediaSession.setCallback(new PlayerSessionCallback());
        mMediaSession.setActive(true);
    }

    private void showMediaNotification(PlaybackStateCompat state) {
        boolean isPlaying = state.getState() == PlaybackStateCompat.STATE_PLAYING;

        int icon = isPlaying ? R.drawable.ic_pause_36dp : R.drawable.ic_play_36dp;
        String playPause = isPlaying ? getString(R.string.pause) : getString(R.string.play);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(this,
                getString(R.string.notification_channel_id));

        NotificationCompat.Action playPauseAction = new NotificationCompat.Action(
                icon, playPause,
                MediaButtonReceiver.buildMediaButtonPendingIntent(this,
                        PlaybackStateCompat.ACTION_PLAY_PAUSE));

        PendingIntent contentPendingIntent = PendingIntent.getActivity(
                this, 0, new Intent(this, PlayerActivity.class), 0);

        builder.setContentIntent(contentPendingIntent)
                .setSmallIcon(R.drawable.ic_youtube_24dp)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .addAction(playPauseAction)
                .setOngoing(isPlaying)
                .setStyle(new android.support.v4.media.app.NotificationCompat.MediaStyle()
                        .setShowActionsInCompactView(0)
                        .setMediaSession(mMediaSession.getSessionToken()));

        if (mVideoThumbnail != null) {
            builder.setContentTitle(mVideoData.getSnippet().getTitle())
                    .setContentText(mVideoData.getSnippet().getChannelTitle())
                    .setLargeIcon(mVideoThumbnail);
        }

        if (mVideoData == null) {
            Log.d(TAG, "Media notification hidden due to lack of the YouTube API data");
            ((NotificationManager) getSystemService(NOTIFICATION_SERVICE)).cancel(0);
            return;
        }

        NotificationManager notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        notificationManager.notify(0, builder.build());
    }

    /*//////////////////////////////////////////////////////////////////////////////
    // Utils
    //////////////////////////////////////////////////////////////////////////////*/

    private String getIntentVideoId() {
        String videoUrl = null;
        if (getIntent() != null && getIntent().getExtras() != null)
            videoUrl = getIntent().getExtras().getString(Intent.EXTRA_TEXT);

        String videoId;

        Log.d(TAG, "Video url: " + videoUrl);

        // Activity started by a share Intent with a video url
        if (videoUrl != null && !videoUrl.equals(""))
            videoId = videoUrl.substring(videoUrl.length() - 11, videoUrl.length());
        // Activity started by a regular Intent with a video id
        else if (getIntent().getExtras().containsKey(getString(R.string.video_id_key)))
            videoId = getIntent().getExtras().getString(getString(R.string.video_id_key));
        // No video to play, throw an exception
        else {
            throw new IllegalArgumentException("No video id available, cannot initialise the player");
        }

        Log.d(TAG, "Video id: " + videoId);

        return videoId;
    }

    /*//////////////////////////////////////////////////////////////////////////////
    // Callbacks and others
    //////////////////////////////////////////////////////////////////////////////*/

    // TODO: replace it with a call to picasso
    private class SetAlbumArtTask extends AsyncTask<String, Void, Bitmap> {
        @Override
        protected Bitmap doInBackground(String... strings) {
            Bitmap result = null;

            try {
                URL url = new URL(strings[0]);
                result = BitmapFactory.decodeStream(url.openConnection().getInputStream());
            } catch (IOException e) {
                Log.e(TAG, e.toString());
            }

            return result;
        }

        @Override
        protected void onPostExecute(Bitmap bitmap) {
            mVideoThumbnail = bitmap;
            showMediaNotification(mStateBuilder.build());
        }
    }

    private class PlayerSessionCallback extends MediaSessionCompat.Callback {
        @Override
        public void onPlay() {
            mainPlayerView.togglePlayPause();
        }

        @Override
        public void onPause() {
            mainPlayerView.togglePlayPause();
        }
    }

    public static class MediaReceiver extends BroadcastReceiver {

        public MediaReceiver() {
        }

        @Override
        public void onReceive(Context context, Intent intent) {
            MediaButtonReceiver.handleIntent(mMediaSession, intent);
        }
    }

    // A service that calls necessary cleanup methods after the player is closed
    public static class PlayerLifecycleService extends Service {

        public PlayerLifecycleService() {
            super();
        }

        @Nullable
        @Override
        public IBinder onBind(Intent intent) {
            return null;
        }

        @Override
        public void onTaskRemoved(Intent rootIntent) {
            super.onTaskRemoved(rootIntent);

            ((NotificationManager) getSystemService(NOTIFICATION_SERVICE)).cancel(0);

            stopSelf();
        }
    }

    @Override
    public void onVideoDataReceived(Video videoData) {
        mVideoData = videoData;

        if (mVideoData.getSnippet().getThumbnails().getStandard() != null)
            new SetAlbumArtTask().execute(mVideoData.getSnippet().getThumbnails().getStandard().getUrl());
    }

    @Override
    public void onSeekButtonClicked(float duration) {
        mainPlayerView.getPlayer().seekTo(duration);
        mainPlayerView.resumePlayback();
    }
}

File: app/src/main/java/com/cajetan/youtubeplus/PlayerActivity.kt
package com.cajetan.youtubeplus

import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import androidx.lifecycle.ViewModelProviders
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.ActivityInfo
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.opengl.Visibility
import android.os.Bundle
import android.os.IBinder
import android.preference.PreferenceManager
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import androidx.appcompat.app.AppCompatActivity
import android.view.View
import android.widget.Toast
import androidx.core.app.NotificationCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.cajetan.youtubeplus.adapters.VideoListAdapter
import com.cajetan.youtubeplus.data.VideoData
import com.cajetan.youtubeplus.data.VideoDataViewModel
import com.cajetan.youtubeplus.fragments.SeekDialogFragment
import com.cajetan.youtubeplus.utils.FullScreenHelper
import com.cajetan.youtubeplus.utils.YouTubeData
import com.google.api.services.youtube.model.PlaylistItem
import com.google.api.services.youtube.model.Video
import com.pierfrancescosoffritti.androidyoutubeplayer.player.PlayerConstants
import com.pierfrancescosoffritti.androidyoutubeplayer.player.listeners.AbstractYouTubePlayerListener
import com.pierfrancescosoffritti.androidyoutubeplayer.player.listeners.YouTubePlayerFullScreenListener
import com.pierfrancescosoffritti.androidyoutubeplayer.ui.PlayerUIController
import com.pierfrancescosoffritti.androidyoutubeplayer.ui.menu.MenuItem
import com.pierfrancescosoffritti.androidyoutubeplayer.ui.menu.YouTubePlayerMenu
import com.pierfrancescosoffritti.androidyoutubeplayer.utils.YouTubePlayerTracker
import kotlinx.android.synthetic.main.activity_player.*
import org.jetbrains.anko.*
import java.lang.IllegalArgumentException
import java.net.URL

class PlayerActivity : AppCompatActivity(), YouTubeData.VideoDataListener,
        YouTubeData.VideoListDataListener, VideoListAdapter.ListItemClickListener,
        SeekDialogFragment.SeekDialogListener, YouTubeData.PlaylistDataListener {

    private val TAG: String = this.javaClass.simpleName

    private val mFullScreenHelper: FullScreenHelper = FullScreenHelper(this)

    private lateinit var mMediaSession: MediaSessionCompat
    private lateinit var mStateBuilder: PlaybackStateCompat.Builder

    private lateinit var mYouTubeData: YouTubeData
    private lateinit var mUIController: PlayerUIController

    private lateinit var mVideoId: String
    private var mVideoData: Video? = null
    private lateinit var mVideoThumbnail: Bitmap

    private val mContext: Context = this
    private lateinit var mVideoDataViewModel: VideoDataViewModel

    private val mTracker: YouTubePlayerTracker = YouTubePlayerTracker()

    private val mBroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.extras?.getString(getString(R.string.player_action_key)) ==
                    getString(R.string.player_action_play_pause)) {
                mainPlayerView?.togglePlayPause()
            }
        }
    }

    // TODO: refactor some methods with default arguments

    private var playlistMode = false
    private lateinit var mAdapter: VideoListAdapter
    private var mPrevPageToken: String = ""
    private var mNextPageToken: String = ""
    private var mCurrentVideoIndex = 0

    ////////////////////////////////////////////////////////////////////////////////
    // Lifecycle
    ////////////////////////////////////////////////////////////////////////////////

    override fun onCreate(savedInstanceState: Bundle?) {
        val darkMode = PreferenceManager.getDefaultSharedPreferences(this)
                .getBoolean("dark_mode", false)
        setTheme(if (darkMode) R.style.PlayerActivityThemeDark else R.style.PlayerActivityThemeLight)

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_player)

        mYouTubeData = YouTubeData(this)
        mVideoDataViewModel = ViewModelProviders.of(this).get(VideoDataViewModel::class.java)

        val playlistId = getPlaylistId(intent)
        when (playlistId) {
            null -> {
                // A single video
                mVideoId = getIntentVideoId(intent!!)
                mYouTubeData.receiveVideoData(mVideoId)
                setupPlayer()
            }
            else -> {
                // A playlist, setup player after receiving the data
                setupPlayer()
                setupPlaylist(playlistId)
            }
        }

        setupMediaSession()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        mPrevPageToken = ""
        mNextPageToken = ""
        mCurrentVideoIndex = 0

        val playlistId = getPlaylistId(intent)
        when (playlistId) {
            null -> {
                playlistMode = false
                // TODO: possible improvements
                if (this::mAdapter.isInitialized) {
                    mAdapter.clearItems()
                    videoList.visibility = View.GONE
                }
                switchVideo(getIntentVideoId(intent))
            }
            else -> {
                playlistMode = true
                if (!this::mAdapter.isInitialized) {
                    setupPlaylist(playlistId)
                } else {
                    mAdapter.switchNowPlaying(0)
                    mAdapter.clearItems()
                    mYouTubeData.receivePlaylistResults(playlistId)
                }
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        mYouTubeData.onParentActivityResult(requestCode, resultCode, data)
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        mYouTubeData.receiveVideoData(mVideoId)
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onDestroy() {
        (getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).cancel(0)
        unregisterReceiver(mBroadcastReceiver)
        mainPlayerView.release()
        super.onDestroy()
    }

    ////////////////////////////////////////////////////////////////////////////////
    // Init
    ////////////////////////////////////////////////////////////////////////////////

    private fun setupPlayer() {
        registerReceiver(mBroadcastReceiver,
                IntentFilter().apply { addAction(Intent.ACTION_MEDIA_BUTTON) }
        )

        startService(Intent(this, PlayerLifecycleService::class.java))

        mainPlayerView.enableBackgroundPlayback(true)
        mainPlayerView.initialize({ initialisedYouTubePlayer ->
            initialisedYouTubePlayer.addListener(mTracker)
            initialisedYouTubePlayer.addListener(object: AbstractYouTubePlayerListener() {
                override fun onReady() {
                    if (mVideoId.isNotEmpty())
                        initialisedYouTubePlayer.loadVideo(mVideoId, 0F)
                }

                override fun onStateChange(state: PlayerConstants.PlayerState) {
                    if (playlistMode && state == PlayerConstants.PlayerState.ENDED) {
                        mCurrentVideoIndex++
                        switchVideo(mAdapter.getItem(mCurrentVideoIndex).id)
                        mAdapter.switchNowPlaying(mCurrentVideoIndex)
                    }

                    val playerState: Int? = when (state) {
                        PlayerConstants.PlayerState.PLAYING -> PlaybackStateCompat.STATE_PLAYING
                        PlayerConstants.PlayerState.PAUSED -> PlaybackStateCompat.STATE_PAUSED
                        PlayerConstants.PlayerState.ENDED -> PlaybackStateCompat.STATE_PAUSED
                        else -> null
                    }

                    if (playerState != null)
                        mStateBuilder.setState(playerState, mTracker.currentSecond.toLong(), 1f)

                    mMediaSession.setPlaybackState(mStateBuilder.build())
                    showMediaNotification(mStateBuilder.build())
                    super.onStateChange(state)
                }
            })
        }, true)

        mainPlayerView.addFullScreenListener(object: YouTubePlayerFullScreenListener {
            override fun onYouTubePlayerEnterFullScreen() {
                requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
                mFullScreenHelper.enterFullScreen()
            }

            override fun onYouTubePlayerExitFullScreen() {
                requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
                mFullScreenHelper.exitFullScreen()
            }
        })

        mUIController = mainPlayerView.playerUIController

        if (mUIController.menu == null)
            return

        mUIController.showMenuButton(true)
        addMenuItems(mUIController.menu as YouTubePlayerMenu)
    }

    private fun setupPlaylist(playlistId: String) {
        playlistMode = true
        mAdapter = VideoListAdapter(emptyList(), this, this)
        mAdapter.switchNowPlaying(mCurrentVideoIndex)
        mAdapter.onBottomReached = {
            if (mNextPageToken.isNotEmpty()) {
                progressBarBottom.visibility = View.VISIBLE
                mYouTubeData.receivePlaylistResults(playlistId, mNextPageToken)
            }
        }

        videoList.layoutManager = LinearLayoutManager(this)
        videoList.setHasFixedSize(false)
        videoList.adapter = mAdapter

        mYouTubeData.receivePlaylistResults(playlistId)
    }

    private fun addMenuItems(menu: YouTubePlayerMenu) {
        menu.addItem(MenuItem(getString(R.string.seek),
                R.drawable.ic_timer_black_24dp) {

            val bundle = Bundle().apply {
                putString(getString(R.string.duration_string_key),
                        mVideoData?.contentDetails?.duration)
                putInt(getString(R.string.current_second_key),
                        Math.round(mTracker.currentSecond))
            }

            SeekDialogFragment().apply {
                arguments = bundle
            }.show(supportFragmentManager, getString(R.string.seeker_dialog_id))

            menu.dismiss()
        })

        doAsync {
            val contains = mVideoDataViewModel.contains(mVideoId)

            val text = when (contains) {
                true -> getString(R.string.favourites_remove)
                false -> getString(R.string.favourites_add)
            }

            val icon = when (contains) {
                true -> R.drawable.ic_star_black_24dp
                false -> R.drawable.ic_star_border_black_24dp
            }

            val message = when (contains) {
                true -> getString(R.string.favourites_removed)
                false -> getString(R.string.favourites_added)
            }

            uiThread {
                menu.addItem(MenuItem(text, icon) {
                    when (contains) {
                        true -> mVideoDataViewModel.delete(VideoData(mVideoId))
                        false -> mVideoDataViewModel.insert(VideoData(mVideoId))
                    }

                    Toast.makeText(mContext, message, Toast.LENGTH_SHORT).show()

                    // Remove menu items
                    for (x in 0 until menu.itemCount)
                        menu.removeItem(0)

                    // Add items again to update the state
                    addMenuItems(menu)
                    menu.dismiss()
                })
            }
        }
    }

    private fun setupMediaSession() {
        mMediaSession = MediaSessionCompat(this, TAG)

        mMediaSession.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS or
                MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS)

        mMediaSession.setMediaButtonReceiver(null)
        mStateBuilder = PlaybackStateCompat.Builder().setActions(
                PlaybackStateCompat.ACTION_PLAY or
                        PlaybackStateCompat.ACTION_PAUSE or
//                        PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS or
                        PlaybackStateCompat.ACTION_PLAY_PAUSE)

        mMediaSession.setPlaybackState(mStateBuilder.build())
        mMediaSession.setCallback(PlayerSessionCallback())
        mMediaSession.isActive = true
    }

    private fun showMediaNotification(state: PlaybackStateCompat) {
        val isPlaying = state.state == PlaybackStateCompat.STATE_PLAYING

        val icon = if (isPlaying) R.drawable.ic_pause_36dp else R.drawable.ic_play_36dp
        val playPause = if (isPlaying) getString(R.string.pause) else getString(R.string.play)

        val builder: NotificationCompat.Builder = NotificationCompat.Builder(this,
                getString(R.string.notification_channel_id))

        val playPauseAction = NotificationCompat.Action(icon, playPause,
                PendingIntent.getBroadcast(this, 2923588,
                        Intent().apply {
                            action = Intent.ACTION_MEDIA_BUTTON
                            putExtra(getString(R.string.player_action_key),
                                    getString(R.string.player_action_play_pause))
                        },
                        PendingIntent.FLAG_UPDATE_CURRENT))

        val contentPendingIntent = PendingIntent.getActivity(this, 0,
                Intent(this, PlayerActivity::class.java), 0)

        builder.setContentIntent(contentPendingIntent)
                .setSmallIcon(R.drawable.ic_youtube_24dp)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .addAction(playPauseAction)
                .setOngoing(isPlaying)
                .setStyle(androidx.media.app.NotificationCompat.MediaStyle()
                        .setShowActionsInCompactView(0)
                        .setMediaSession(mMediaSession.sessionToken))

        if (this::mVideoThumbnail.isInitialized) {
            builder.setContentTitle(mVideoData?.snippet?.title)
                    .setContentText(mVideoData?.snippet?.channelTitle)
                    .setLargeIcon(mVideoThumbnail)
        }

        if (mVideoData == null) {
            (getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).cancel(0)
            return
        }

        (getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).notify(0, builder.build())
    }

    ////////////////////////////////////////////////////////////////////////////////
    // Utils
    ////////////////////////////////////////////////////////////////////////////////

    private fun getIntentVideoId(intent: Intent): String {
        val extras: Bundle = intent.extras!!
        val videoUrl: String = extras.getString(Intent.EXTRA_TEXT) ?: ""

        return when {
            videoUrl.isNotEmpty() -> videoUrl.substring(videoUrl.length - 11, videoUrl.length)

            extras.containsKey(getString(R.string.video_id_key)) ->
                extras.getString(getString(R.string.video_id_key)) as String

            else -> throw IllegalArgumentException("No video id available, cannot initialise the player")
        }
    }

    private fun getPlaylistId(intent: Intent?): String? {
        val videoUrl = intent?.extras?.getString(Intent.EXTRA_TEXT) ?: return null

        if (!videoUrl.contains("playlist?list="))
            return null

        return videoUrl.substring(videoUrl.indexOf("playlist?list=", 0) + 14)
    }

    private fun switchVideo(videoId: String) {
        mVideoId = videoId
        mVideoData = null
        mYouTubeData.receiveVideoData(videoId)
        mainPlayerView.player.loadVideo(videoId, 0f)
    }

    ////////////////////////////////////////////////////////////////////////////////
    // Callbacks & others
    ////////////////////////////////////////////////////////////////////////////////

    private fun setAlbumArt(url: String) {
        doAsync {
            val bitmap = BitmapFactory.decodeStream(URL(url).openConnection().getInputStream())

            uiThread {
                mVideoThumbnail = bitmap
                showMediaNotification(mStateBuilder.build())
            }
        }
    }

    private inner class PlayerSessionCallback : MediaSessionCompat.Callback() {
        override fun onPlay() {
            mainPlayerView.togglePlayPause()
        }

        override fun onPause() {
            mainPlayerView.togglePlayPause()
        }
    }

    class PlayerLifecycleService : Service() {
        override fun onBind(intent: Intent?): IBinder? {
            return null
        }

        override fun onTaskRemoved(rootIntent: Intent?) {
            super.onTaskRemoved(rootIntent)

            (getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).cancel(0)
            stopSelf()
        }
    }

    override fun onVideoDataReceived(videoData: Video) {
        mVideoData = videoData

        if (mVideoData?.snippet?.thumbnails?.standard != null) {
            val url = mVideoData?.snippet?.thumbnails?.standard?.url
            if (url != null)
                setAlbumArt(url)
        }

        if (mVideoData?.contentDetails?.duration == getString(R.string.live_video_duration))
            mainPlayerView.playerUIController.enableLiveVideoUI(true)
    }

    override fun onPlaylistDataReceived(results: List<PlaylistItem>,
                                        nextPageToken: String, previousPageToken: String) {
        // First call
        if (previousPageToken.isEmpty())
            switchVideo(results[0].contentDetails.videoId)

        val playlistItems: List<VideoData> = results.map { VideoData(it.contentDetails.videoId) }

        mPrevPageToken = previousPageToken
        mNextPageToken = nextPageToken
        mYouTubeData.receiveVideoListResults(playlistItems)
    }

    override fun onVideoListReceived(results: List<Video>, block: ((List<Video>) -> List<Video>)?) {
        val result = block?.invoke(results)?.toList() ?: results.toList()
        mAdapter.addItems(result)

        videoList.visibility = View.VISIBLE
        progressBarBottom.visibility = View.GONE
    }

    override fun onListItemClick(clickedVideoId: String, position: Int) {
        switchVideo(clickedVideoId)
        mAdapter.switchNowPlaying(position)
    }

    override fun onListItemLongClick(clickedVideoId: String) {
        this.alert(getString(R.string.favourite_add_confirmation)) {
            yesButton { mVideoDataViewModel.insert(VideoData(clickedVideoId)) }
            noButton { }
        }.show()
    }

    override fun onSeekButtonClicked(duration: Float) {
        mainPlayerView.player.seekTo(duration)
        mainPlayerView.resumePlayback()
    }
}
